src/Features/Core/Portable/ConvertTupleToStruct/AbstractConvertTupleToStructCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,865 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.GenerateEqualsAndGetHashCodeFromMembers;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ConvertTupleToStruct
+{
+    internal abstract partial class AbstractConvertTupleToStructCodeRefactoringProvider<
+        TExpressionSyntax,
+        TNameSyntax,
+        TIdentifierNameSyntax,
+        TLiteralExpressionSyntax,
+        TObjectCreationExpressionSyntax,
+        TTupleExpressionSyntax,
+        TArgumentSyntax,
+        TTupleTypeSyntax,
+        TTypeBlockSyntax,
+        TNamespaceDeclarationSyntax>
+        : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TNameSyntax : TExpressionSyntax
+        where TIdentifierNameSyntax : TNameSyntax
+        where TLiteralExpressionSyntax : TExpressionSyntax
+        where TObjectCreationExpressionSyntax : TExpressionSyntax
+        where TTupleExpressionSyntax : TExpressionSyntax
+        where TArgumentSyntax: SyntaxNode
+        where TTupleTypeSyntax : SyntaxNode
+        where TTypeBlockSyntax : SyntaxNode
+        where TNamespaceDeclarationSyntax : SyntaxNode
+    {
+        private enum Scope
+        {
+            ContainingMember,
+            ContainingType,
+            ContainingProject,
+            DependentProjects
+        }
+
+        protected abstract TObjectCreationExpressionSyntax CreateObjectCreationExpression(
+            TNameSyntax nameNode, SyntaxToken openParen, SeparatedSyntaxList<TArgumentSyntax> arguments, SyntaxToken closeParen);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var (tupleExprOrTypeNode, tupleType) = await TryGetTupleInfoAsync(
+                document, context.Span, cancellationToken).ConfigureAwait(false);
+
+            if (tupleExprOrTypeNode == null || tupleType == null)
+            {
+                return;
+            }
+
+            // Check if the anonymous type actually references another anonymous type inside of it.
+            // If it does, we can't convert this.  There is no way to describe this anonymous type
+            // in the concrete type we create.
+            var fields = tupleType.TupleElements;
```

Comment:
>fields [](start = 16, length = 6)

Nit: consider renaming to `elements`

Reply:
i personally like 'fields' because it makes it clear to me that i have a collection of IFieldSymbols.  'elements' is actually a little too vague for me :)